client/my-sites/sharing/connections/service.jsx

```diff
@@ -1,240 +1,378 @@
+	 */
+	addConnection = ( service, keyringConnectionId, externalUserId = 0 ) => {
+		this.setState( { isConnecting: true } );
 
-	addConnection: function( service, keyringConnectionId, externalUserId ) {
 		if ( service ) {
-			// Attempt to create a new connection. If a Keyring connection ID
-			// is not provided, the user will need to authorize the app
-			this.props.connections.create( service, this.props.siteId, keyringConnectionId, externalUserId );
-
-			// In the case that a Keyring connection doesn't exist, wait for app
-			// authorization to occur, then display with the available connections
-			if ( ! keyringConnectionId ) {
-				this.props.connections.once( 'connect', function() {
-					if ( serviceConnections.didKeyringConnectionSucceed( service.ID, this.props.siteId ) &&
-						serviceConnections.isServiceForPublicize( service.ID ) ) {
-						this.setState( { isSelectingAccount: true } );
-					}
-				}.bind( this ) );
-			} else {
+			if ( keyringConnectionId ) {
+				// Since we have a Keyring connection to work with, we can immediately
+				// create or update the connection
+				const existingConnection = find( this.props.siteUserConnections, { keyring_connection_ID: keyringConnectionId } );
+
+				if ( this.props.siteId && existingConnection ) {
+					// If a Keyring connection is already in use by another connection,
+					// we should trigger an update. There should only be one connection,
+					// so we're correct in using the connection ID from the first
+					this.props.updateSiteConnection( existingConnection, { external_user_ID: externalUserId } );
+				} else {
+					this.props.createSiteConnection( this.props.siteId, keyringConnectionId, externalUserId );
+				}
+
 				this.props.recordGoogleEvent( 'Sharing', 'Clicked Connect Button in Modal', this.props.service.ID );
+			} else {
+				// Attempt to create a new connection. If a Keyring connection ID
+				// is not provided, the user will need to authorize the app
+				const popupMonitor = new PopupMonitor();
+
+				popupMonitor.open( service.connect_URL, null, 'toolbar=0,location=0,status=0,menubar=0,' +
+					popupMonitor.getScreenCenterSpecs( 780, 500 ) );
+
+				popupMonitor.once( 'close', () => {
+					// When the user has finished authorizing the connection
+					// (or otherwise closed the window), force a refresh
+					this.props.requestKeyringConnections();
+
+					// In the case that a Keyring connection doesn't exist, wait for app
+					// authorization to occur, then display with the available connections
+					this.setState( { isAwaitingConnections: true } );
+				} );
 			}
 		} else {
 			// If an account wasn't selected from the dialog or the user cancels
 			// the connection, the dialog should simply close
-			this.props.connections.emit( 'create:error', { cancel: true } );
+			this.props.errorNotice( this.props.translate( 'The connection could not be made because no account was selected.', {
```

Comment:
Alternate string suggestion: 
* `translate( 'The %(service)s connection could not be made because no account was selected.', { context: 'Sharing: Publicize connection confirmation'} )`&mdash; translations: **21**. 

[PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/remove/connections-list-module)